{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d72edae6-95d4-4e9d-83df-2557707b2d7e",
   "metadata": {},
   "source": [
    "### WGKB Gene Database Query"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0e708f1",
   "metadata": {},
   "source": [
    "#### Reading data via web scraping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed07bfae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "811ebb2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_001411555.2/'\n",
    "\n",
    "payload = {\n",
    "    'accession': 'GCF_001411555.2',\n",
    "    'gene_types': [],\n",
    "    'locations': [],\n",
    "    'page_size': 20,\n",
    "    'page_token': '',\n",
    "    'search_text': [],\n",
    "    'sort': []\n",
    "}\n",
    "\n",
    "headers = {\n",
    "    'Origin': 'https://www.ncbi.nlm.nih.gov',\n",
    "    'Referer': 'https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_001411555.2/',\n",
    "    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36'\n",
    "}\n",
    "\n",
    "result = requests.post(url, headers=headers, params=payload)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "177bf931",
   "metadata": {},
   "source": [
    "#### Reading data via REST API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6fe5b161-b65e-4126-8fba-5a7a90d9a750",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "debded06-f2b8-407f-b86c-c7a00b7cf9d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://api.ncbi.nlm.nih.gov/datasets/v2alpha'\n",
    "\n",
    "taxon = 'Juglans regia (English walnut)'\n",
    "api_key = 'removed'\n",
    "\n",
    "params = {\n",
    "    'name': taxon,\n",
    "    'api_key': api_key\n",
    "}\n",
    "\n",
    "response = requests.get(url, params=params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "73e5da2f-5f6c-4e54-8f42-90ec9b21959b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'404 page not found\\n'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.text\n",
    "\n",
    "# Might have to enable the NCBI API key. Don't think that will be viable if it's tied to my account."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d316629-d3c8-4c3a-8f98-714688928c61",
   "metadata": {},
   "source": [
    "#### Reading metadata with NCBI documented APIs (GenomeAPI, , GeneAPI, APIClient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b4c07ec-0b49-438d-bfe8-539b69825c46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install ncbi-datasets-pylib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "196fba1e-d923-49cb-a399-dc8bef33566d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ncbi.datasets import GenomeApi, GeneApi\n",
    "from ncbi.datasets.openapi import ApiClient"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f142b9c",
   "metadata": {},
   "source": [
    "##### Get chromosome metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0ec9c70-2936-4062-9eff-cca5aefcee80",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_chrom_locations(organism_name: str) -> list:\n",
    "    client = ApiClient()\n",
    "    genome_api = GenomeApi(client)\n",
    "\n",
    "    # Search for genome assemblies by organism name\n",
    "    genome_summaries = genome_api.assembly_descriptors_by_taxon(organism_name)\n",
    "    genomic_locations = []\n",
    "\n",
    "    if genome_summaries.assemblies:\n",
    "        for chrom_dict in genome_summaries.assemblies[0]['assembly']['chromosomes']:\n",
    "            genomic_locations.append({\n",
    "                'RefSeq': chrom_dict.accession_version,\n",
    "                'Size': chrom_dict.length,\n",
    "                'Chromosome': chrom_dict.name\n",
    "            })\n",
    "    elif (not genome_summaries.assemblies) or (not genomic_locations):\n",
    "        print(f\"No genomes found for organism: {organism_name}\")\n",
    "\n",
    "    return genomic_locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "37efb91d-6801-4593-aff8-452ca594262f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'RefSeq': 'NC_049901.1', 'Size': '45207397', 'Chromosome': '1'},\n",
       " {'RefSeq': 'NC_049902.1', 'Size': '37821870', 'Chromosome': '2'},\n",
       " {'RefSeq': 'NC_049903.1', 'Size': '35064427', 'Chromosome': '3'},\n",
       " {'RefSeq': 'NC_049904.1', 'Size': '34823025', 'Chromosome': '4'},\n",
       " {'RefSeq': 'NC_049905.1', 'Size': '22562875', 'Chromosome': '5'},\n",
       " {'RefSeq': 'NC_049906.1', 'Size': '39020271', 'Chromosome': '6'},\n",
       " {'RefSeq': 'NC_049907.1', 'Size': '52418484', 'Chromosome': '7'},\n",
       " {'RefSeq': 'NC_049908.1', 'Size': '30564197', 'Chromosome': '8'},\n",
       " {'RefSeq': 'NC_049909.1', 'Size': '24263475', 'Chromosome': '9'},\n",
       " {'RefSeq': 'NC_049910.1', 'Size': '37707155', 'Chromosome': '10'},\n",
       " {'RefSeq': 'NC_049911.1', 'Size': '37114715', 'Chromosome': '11'},\n",
       " {'RefSeq': 'NC_049912.1', 'Size': '31492331', 'Chromosome': '12'},\n",
       " {'RefSeq': 'NC_049913.1', 'Size': '39757759', 'Chromosome': '13'},\n",
       " {'RefSeq': 'NC_049914.1', 'Size': '28841373', 'Chromosome': '14'},\n",
       " {'RefSeq': 'NC_049915.1', 'Size': '20407330', 'Chromosome': '15'},\n",
       " {'RefSeq': 'NC_049916.1', 'Size': '28711772', 'Chromosome': '16'},\n",
       " {'RefSeq': 'NC_028617.1', 'Size': '160537', 'Chromosome': 'Pltd'},\n",
       " {'RefSeq': None, 'Size': '27007677', 'Chromosome': 'Un'}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Walnut\n",
    "\n",
    "get_chrom_locations('Juglans regia (English walnut)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14c324a",
   "metadata": {},
   "source": [
    "##### GeneAPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5126461d-fff4-408a-95c8-1f3ec9fdeab3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on GeneApi in module ncbi.datasets.openapi.api.gene_api object:\n",
      "\n",
      "class GeneApi(builtins.object)\n",
      " |  GeneApi(api_client=None)\n",
      " |  \n",
      " |  NOTE: This class is auto generated by OpenAPI Generator\n",
      " |  Ref: https://openapi-generator.tech\n",
      " |  \n",
      " |  Do not edit the class manually.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, api_client=None)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  download_gene_package(self, gene_ids, **kwargs)\n",
      " |      Get a gene dataset by gene ID  # noqa: E501\n",
      " |      \n",
      " |      Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by gene ID.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.download_gene_package(gene_ids, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          gene_ids ([int]): NCBI gene ids\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          include_annotation_type ([V1Fasta]): Select additional types of annotation to include in the data package.  If unset, no annotation is provided.. [optional]\n",
      " |          fasta_filter ([str]): Limit the FASTA sequences in the datasets package to these transcript and protein accessions. [optional]\n",
      " |          filename (str): Output file name.. [optional] if omitted the server will use the default value of \"ncbi_dataset.zip\"\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          file_type\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  download_gene_package_post(self, v1_gene_dataset_request, **kwargs)\n",
      " |      Get a gene dataset by POST  # noqa: E501\n",
      " |      \n",
      " |      Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by POST.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.download_gene_package_post(v1_gene_dataset_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_gene_dataset_request (V1GeneDatasetRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          filename (str): Output file name.. [optional] if omitted the server will use the default value of \"ncbi_dataset.zip\"\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          file_type\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_download_summary_by_accession(self, accessions, **kwargs)\n",
      " |      Get gene download summary by RefSeq Accession  # noqa: E501\n",
      " |      \n",
      " |      Get gene download summary by RefSeq Accession in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_download_summary_by_accession(accessions, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          accessions ([str]): RNA or Protein accessions.\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          fasta_filter ([str]): Limit the FASTA sequences in the datasets package to these transcript and protein accessions. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1DownloadSummary\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_download_summary_by_id(self, gene_ids, **kwargs)\n",
      " |      Get gene download summary by GeneID  # noqa: E501\n",
      " |      \n",
      " |      Get a download summary by GeneID in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_download_summary_by_id(gene_ids, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          gene_ids ([int]): NCBI gene ids\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          fasta_filter ([str]): Limit the FASTA sequences in the datasets package to these transcript and protein accessions. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1DownloadSummary\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_download_summary_by_post(self, v1_gene_dataset_request, **kwargs)\n",
      " |      Get gene download summary  # noqa: E501\n",
      " |      \n",
      " |      Get gene download summary in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_download_summary_by_post(v1_gene_dataset_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_gene_dataset_request (V1GeneDatasetRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1DownloadSummary\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_download_summary_by_tax_and_symbol(self, symbols, taxon, **kwargs)\n",
      " |      Get gene download summary by gene symbol  # noqa: E501\n",
      " |      \n",
      " |      Get gene download summary by gene symbol in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_download_summary_by_tax_and_symbol(symbols, taxon, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          symbols ([str]): Gene symbol\n",
      " |          taxon (str): Taxon for provided gene symbol\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          fasta_filter ([str]): Limit the FASTA sequences in the datasets package to these transcript and protein accessions. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1DownloadSummary\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_metadata_by_accession(self, accessions, **kwargs)\n",
      " |      Get gene metadata by RefSeq Accession  # noqa: E501\n",
      " |      \n",
      " |      Get detailed gene metadata by RefSeq Accession in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_metadata_by_accession(accessions, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          accessions ([str]): RNA or Protein accessions.\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          returned_content (V1GeneDatasetRequestContentType): Return either gene-ids, or entire gene metadata. [optional]\n",
      " |          sort_schema_field (V1GeneDatasetRequestSortField): Select a field to sort on.. [optional]\n",
      " |          sort_schema_direction (V1SortDirection): Select a direction for the sort.. [optional]\n",
      " |          limit (str): Limit the number of returned results (\"all\", \"none\", otherwise an integer value). [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1GeneMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_metadata_by_id(self, gene_ids, **kwargs)\n",
      " |      Get gene metadata by GeneID  # noqa: E501\n",
      " |      \n",
      " |      Get detailed gene metadata by GeneID in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_metadata_by_id(gene_ids, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          gene_ids ([int]): NCBI gene ids\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          returned_content (V1GeneDatasetRequestContentType): Return either gene-ids, or entire gene metadata. [optional]\n",
      " |          sort_schema_field (V1GeneDatasetRequestSortField): Select a field to sort on.. [optional]\n",
      " |          sort_schema_direction (V1SortDirection): Select a direction for the sort.. [optional]\n",
      " |          limit (str): Limit the number of returned results (\"all\", \"none\", otherwise an integer value). [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1GeneMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_metadata_by_post(self, v1_gene_dataset_request, **kwargs)\n",
      " |      Get gene metadata as JSON  # noqa: E501\n",
      " |      \n",
      " |      Get detailed gene metadata in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_metadata_by_post(v1_gene_dataset_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_gene_dataset_request (V1GeneDatasetRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1GeneMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_metadata_by_tax_and_symbol(self, symbols, taxon, **kwargs)\n",
      " |      Get gene metadata by gene symbol  # noqa: E501\n",
      " |      \n",
      " |      Get detailed gene metadata by gene symbol in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_metadata_by_tax_and_symbol(symbols, taxon, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          symbols ([str]): Gene symbol\n",
      " |          taxon (str): Taxon for provided gene symbol\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          accessions ([str]): RNA or Protein accessions.. [optional]\n",
      " |          returned_content (V1GeneDatasetRequestContentType): Return either gene-ids, or entire gene metadata. [optional]\n",
      " |          sort_schema_field (V1GeneDatasetRequestSortField): Select a field to sort on.. [optional]\n",
      " |          sort_schema_direction (V1SortDirection): Select a direction for the sort.. [optional]\n",
      " |          limit (str): Limit the number of returned results (\"all\", \"none\", otherwise an integer value). [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1GeneMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_metadata_stream_by_post(self, v1_gene_dataset_request, **kwargs)\n",
      " |      Get gene metadata  # noqa: E501\n",
      " |      \n",
      " |      Get detailed gene metadata in a streaming, JSON-lines output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_metadata_stream_by_post(v1_gene_dataset_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_gene_dataset_request (V1GeneDatasetRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1GeneMatch\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_orthologs_by_id(self, gene_id, **kwargs)\n",
      " |      Get gene orthologs by gene ID  # noqa: E501\n",
      " |      \n",
      " |      Get detailed gene metadata for an ortholog set by gene ID in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_orthologs_by_id(gene_id, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          gene_id (int):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          returned_content (V1OrthologRequestContentType): Return either gene-ids, or entire gene metadata. [optional]\n",
      " |          taxon_filter ([str]): Filter genes by taxa. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1OrthologSet\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_tax_name_query(self, taxon_query, **kwargs)\n",
      " |      Get a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name  # noqa: E501\n",
      " |      \n",
      " |      This endpoint retrieves a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name of any rank.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_tax_name_query(taxon_query, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          taxon_query (str): NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          tax_rank_filter (V1OrganismQueryRequestTaxRankFilter): Set the scope of searched tax ranks. [optional]\n",
      " |          tax_resource_filter (V1OrganismQueryRequestTaxonResourceFilter): [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1SciNameAndIds\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  gene_tax_tree(self, taxon, **kwargs)\n",
      " |      Get a taxonomic subtree by taxonomic identifier  # noqa: E501\n",
      " |      \n",
      " |      Using an NCBI Taxonomy ID or name (common or scientific) at any rank, get a subtree filtered for species with assembled genomes.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.gene_tax_tree(taxon, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          taxon (str): NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          children_only (bool): Only report the children of the requested taxon and not their descendants. [optional] if omitted the server will use the default value of False\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1Organism\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gene_api = GeneApi()\n",
    "help(gene_api)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ed728c51-bbd8-4c82-beaf-b7ea1f03a30a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sci_name_and_ids': [{'common_name': 'English walnut',\n",
       "                       'sci_name': 'Juglans regia',\n",
       "                       'tax_id': '51240'},\n",
       "                      {'sci_name': 'Juglans microcarpa x Juglans regia',\n",
       "                       'tax_id': '2249226'},\n",
       "                      {'sci_name': 'Juglans sigillata x Juglans regia',\n",
       "                       'tax_id': '1441050'},\n",
       "                      {'sci_name': 'Juglans major x Juglans regia',\n",
       "                       'tax_id': '1141591'},\n",
       "                      {'sci_name': 'Roystonea regia', 'tax_id': '145709'},\n",
       "                      {'sci_name': 'Drosera regia', 'tax_id': '4371'},\n",
       "                      {'sci_name': 'Odontella regia', 'tax_id': '1335017'},\n",
       "                      {'common_name': 'English oak',\n",
       "                       'sci_name': 'Quercus robur',\n",
       "                       'tax_id': '38942'},\n",
       "                      {'common_name': 'black walnut',\n",
       "                       'sci_name': 'Juglans nigra',\n",
       "                       'tax_id': '16719'},\n",
       "                      {'sci_name': 'Delonix regia', 'tax_id': '72433'},\n",
       "                      {'common_name': 'English yew',\n",
       "                       'sci_name': 'Taxus baccata',\n",
       "                       'tax_id': '25629'},\n",
       "                      {'common_name': 'English grain aphid',\n",
       "                       'sci_name': 'Sitobion avenae',\n",
       "                       'tax_id': '44664'},\n",
       "                      {'common_name': 'sea walnut',\n",
       "                       'sci_name': 'Mnemiopsis leidyi',\n",
       "                       'tax_id': '27923'},\n",
       "                      {'common_name': 'English sole',\n",
       "                       'sci_name': 'Parophrys vetulus',\n",
       "                       'tax_id': '172659'},\n",
       "                      {'sci_name': 'Alcidodes juglans', 'tax_id': '2530216'},\n",
       "                      {'sci_name': 'Juglans cathayensis', 'tax_id': '91213'},\n",
       "                      {'sci_name': 'Juglans major', 'tax_id': '91217'},\n",
       "                      {'sci_name': 'Juglans microcarpa', 'tax_id': '91219'}]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "walnut_accessions = get_chrom_locations('Juglans regia (English walnut)')\n",
    "walnut_ref_seq = [chrom['RefSeq'] for chrom in walnut_accessions if chrom['RefSeq'] is not None]\n",
    "# walnut_ref_seq\n",
    "\n",
    "gene_api.gene_tax_name_query('Juglans regia (English walnut)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99cdb7ae",
   "metadata": {},
   "source": [
    "##### GenomeAPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "27c4f540-4fb4-4af7-a53d-b79e90cf5e43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on GenomeApi in module ncbi.datasets.openapi.api.genome_api object:\n",
      "\n",
      "class GenomeApi(builtins.object)\n",
      " |  GenomeApi(api_client=None)\n",
      " |  \n",
      " |  NOTE: This class is auto generated by OpenAPI Generator\n",
      " |  Ref: https://openapi-generator.tech\n",
      " |  \n",
      " |  Do not edit the class manually.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, api_client=None)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  assembly_descriptors_by_accessions(self, accessions, **kwargs)\n",
      " |      Get genome metadata by accession  # noqa: E501\n",
      " |      \n",
      " |      Get detailed metadata for assembled genomes by accession in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.assembly_descriptors_by_accessions(accessions, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          accessions ([str]):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          filters_reference_only (bool): If true, only return reference and representative (GCF_ and GCA_) genome assemblies.. [optional] if omitted the server will use the default value of False\n",
      " |          filters_assembly_source (V1AssemblyDatasetDescriptorsFilterAssemblySource): Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies. [optional]\n",
      " |          filters_has_annotation (bool): Return only annotated genome assemblies. [optional] if omitted the server will use the default value of False\n",
      " |          filters_exclude_atypical (bool): If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical. [optional] if omitted the server will use the default value of False\n",
      " |          filters_assembly_level ([V1AssemblyDatasetDescriptorsFilterAssemblyLevel]): Only return genome assemblies that have one of the specified assembly levels. By default, do not filter.. [optional]\n",
      " |          filters_first_release_date (datetime): Only return genome assemblies that were released on or after the specified date By default, do not filter.. [optional]\n",
      " |          filters_last_release_date (datetime): Only return genome assemblies that were released on or before to the specified date By default, do not filter.. [optional]\n",
      " |          filters_search_text ([str]): Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter. [optional]\n",
      " |          page_size (int): The maximum number of genome assemblies to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results.. [optional] if omitted the server will use the default value of 20\n",
      " |          page_token (str): A page token is returned from an `AssemblyMetadataRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyMetadataRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1AssemblyMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  assembly_descriptors_by_bioproject(self, accessions, **kwargs)\n",
      " |      Get genome metadata by bioproject accession  # noqa: E501\n",
      " |      \n",
      " |      Get detailed metadata for assembled genomes by bioproject accession in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.assembly_descriptors_by_bioproject(accessions, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          accessions ([str]):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          filters_reference_only (bool): If true, only return reference and representative (GCF_ and GCA_) genome assemblies.. [optional] if omitted the server will use the default value of False\n",
      " |          filters_assembly_source (V1AssemblyDatasetDescriptorsFilterAssemblySource): Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies. [optional]\n",
      " |          filters_has_annotation (bool): Return only annotated genome assemblies. [optional] if omitted the server will use the default value of False\n",
      " |          filters_exclude_paired_reports (bool): For paired (GCA/GCF) records, only return the primary record. [optional] if omitted the server will use the default value of False\n",
      " |          filters_exclude_atypical (bool): If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical. [optional] if omitted the server will use the default value of False\n",
      " |          filters_assembly_version (V1AssemblyDatasetDescriptorsFilterAssemblyVersion): Return all assemblies, including replaced and suppressed, or only current assemblies. [optional]\n",
      " |          filters_assembly_level ([V1AssemblyDatasetDescriptorsFilterAssemblyLevel]): Only return genome assemblies that have one of the specified assembly levels. By default, do not filter.. [optional]\n",
      " |          filters_first_release_date (datetime): Only return genome assemblies that were released on or after the specified date By default, do not filter.. [optional]\n",
      " |          filters_last_release_date (datetime): Only return genome assemblies that were released on or before to the specified date By default, do not filter.. [optional]\n",
      " |          filters_search_text ([str]): Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter. [optional]\n",
      " |          returned_content (V1AssemblyMetadataRequestContentType): Return either assembly accessions, or entire assembly-metadata records. [optional]\n",
      " |          page_size (int): The maximum number of genome assemblies to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results.. [optional] if omitted the server will use the default value of 20\n",
      " |          page_token (str): A page token is returned from an `AssemblyMetadataRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyMetadataRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1AssemblyMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  assembly_descriptors_by_taxon(self, taxon, **kwargs)\n",
      " |      Get genome metadata by taxonomic identifier  # noqa: E501\n",
      " |      \n",
      " |      Get detailed metadata on all assembled genomes for a specified NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.assembly_descriptors_by_taxon(taxon, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          taxon (str): NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          filters_reference_only (bool): If true, only return reference and representative (GCF_ and GCA_) genome assemblies.. [optional] if omitted the server will use the default value of False\n",
      " |          filters_assembly_source (V1AssemblyDatasetDescriptorsFilterAssemblySource): Return only RefSeq (GCF_) or GenBank (GCA_) genome assemblies. [optional]\n",
      " |          filters_has_annotation (bool): Return only annotated genome assemblies. [optional] if omitted the server will use the default value of False\n",
      " |          filters_exclude_paired_reports (bool): For paired (GCA/GCF) records, only return the primary record. [optional] if omitted the server will use the default value of False\n",
      " |          filters_exclude_atypical (bool): If true, exclude atypical genomes, i.e. genomes that have assembly issues or are otherwise atypical. [optional] if omitted the server will use the default value of False\n",
      " |          filters_assembly_version (V1AssemblyDatasetDescriptorsFilterAssemblyVersion): Return all assemblies, including replaced and suppressed, or only current assemblies. [optional]\n",
      " |          filters_assembly_level ([V1AssemblyDatasetDescriptorsFilterAssemblyLevel]): Only return genome assemblies that have one of the specified assembly levels. By default, do not filter.. [optional]\n",
      " |          filters_first_release_date (datetime): Only return genome assemblies that were released on or after the specified date By default, do not filter.. [optional]\n",
      " |          filters_last_release_date (datetime): Only return genome assemblies that were released on or before to the specified date By default, do not filter.. [optional]\n",
      " |          filters_search_text ([str]): Only return results whose fields contain the specified search terms in their taxon, infraspecific, assembly name or submitter fields By default, do not filter. [optional]\n",
      " |          tax_exact_match (bool): If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. Ignored for assembly_accession request.. [optional] if omitted the server will use the default value of False\n",
      " |          returned_content (V1AssemblyMetadataRequestContentType): Return either assembly accessions, or entire assembly-metadata records. [optional]\n",
      " |          page_size (int): The maximum number of genome assemblies to return. Default is 20 and maximum is 1000. If the number of results exceeds the page size, `page_token` can be used to retrieve the remaining results.. [optional] if omitted the server will use the default value of 20\n",
      " |          page_token (str): A page token is returned from an `AssemblyMetadataRequest` call with more than `page_size` results. Use this token, along with the previous `AssemblyMetadataRequest` parameters, to retrieve the next page of results. When `page_token` is empty, all results have been retrieved.. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1AssemblyMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  check_assembly_availability(self, accessions, **kwargs)\n",
      " |      Check the validity of genome accessions  # noqa: E501\n",
      " |      \n",
      " |      The 'GET' version of check is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.check_assembly_availability(accessions, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          accessions ([str]): NCBI genome assembly accessions\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1AssemblyDatasetAvailability\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  check_assembly_availability_post(self, v1_assembly_dataset_request, **kwargs)\n",
      " |      Check the validity of many genome accessions in a single request  # noqa: E501\n",
      " |      \n",
      " |      The 'GET' version of check is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.check_assembly_availability_post(v1_assembly_dataset_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_assembly_dataset_request (V1AssemblyDatasetRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1AssemblyDatasetAvailability\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  download_assembly_package(self, accessions, **kwargs)\n",
      " |      Get a genome dataset by accession  # noqa: E501\n",
      " |      \n",
      " |      Download a genome dataset including fasta sequence, annotation and a detailed data report by accession.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.download_assembly_package(accessions, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          accessions ([str]): NCBI genome assembly accessions\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          chromosomes ([str]): The default setting is all chromosome. Specify individual chromosome by string (1,2,MT or chr1,chr2.chrMT). Unplaced sequences are treated like their own chromosome ('Un'). The filter only applies to fasta sequence.. [optional]\n",
      " |          exclude_sequence (bool): Set to true to omit the genomic sequence.. [optional] if omitted the server will use the default value of False\n",
      " |          include_annotation_type ([V1AnnotationForAssemblyType]): Select additional types of annotation to include in the data package.  If unset, no annotation is provided.. [optional]\n",
      " |          hydrated (V1AssemblyDatasetRequestResolution): Set to DATA_REPORT_ONLY, to only retrieve data-reports.. [optional]\n",
      " |          filename (str): Output file name.. [optional] if omitted the server will use the default value of \"ncbi_dataset.zip\"\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          file_type\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  download_assembly_package_post(self, v1_assembly_dataset_request, **kwargs)\n",
      " |      Get a genome dataset by post  # noqa: E501\n",
      " |      \n",
      " |      The 'GET' version of download is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.download_assembly_package_post(v1_assembly_dataset_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_assembly_dataset_request (V1AssemblyDatasetRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          filename (str): Output file name.. [optional] if omitted the server will use the default value of \"ncbi_dataset.zip\"\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          file_type\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  genome_download_summary(self, accessions, **kwargs)\n",
      " |      Preview genome dataset download  # noqa: E501\n",
      " |      \n",
      " |      Get a download summary by accession in a JSON output format.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.genome_download_summary(accessions, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          accessions ([str]): NCBI genome assembly accessions\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          chromosomes ([str]): The default setting is all chromosome. Specify individual chromosome by string (1,2,MT or chr1,chr2.chrMT). Unplaced sequences are treated like their own chromosome ('Un'). The filter only applies to fasta sequence.. [optional]\n",
      " |          exclude_sequence (bool): Set to true to omit the genomic sequence.. [optional] if omitted the server will use the default value of False\n",
      " |          include_annotation_type ([V1AnnotationForAssemblyType]): Select additional types of annotation to include in the data package.  If unset, no annotation is provided.. [optional]\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1DownloadSummary\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  genome_download_summary_by_post(self, v1_assembly_dataset_request, **kwargs)\n",
      " |      Preview genome dataset download by POST  # noqa: E501\n",
      " |      \n",
      " |      The 'GET' version of download summary is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.genome_download_summary_by_post(v1_assembly_dataset_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_assembly_dataset_request (V1AssemblyDatasetRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1DownloadSummary\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  genome_metadata_by_post(self, v1_assembly_metadata_request, **kwargs)\n",
      " |      Get genome metadata by variety of identifiers  # noqa: E501\n",
      " |      \n",
      " |      Get detailed metadata for assembled genomes.  # noqa: E501\n",
      " |      This method makes a synchronous HTTP request by default. To make an\n",
      " |      asynchronous HTTP request, please pass async_req=True\n",
      " |      \n",
      " |      >>> thread = api.genome_metadata_by_post(v1_assembly_metadata_request, async_req=True)\n",
      " |      >>> result = thread.get()\n",
      " |      \n",
      " |      Args:\n",
      " |          v1_assembly_metadata_request (V1AssemblyMetadataRequest):\n",
      " |      \n",
      " |      Keyword Args:\n",
      " |          _return_http_data_only (bool): response data without head status\n",
      " |              code and headers. Default is True.\n",
      " |          _preload_content (bool): if False, the urllib3.HTTPResponse object\n",
      " |              will be returned without reading/decoding response data.\n",
      " |              Default is True.\n",
      " |          _request_timeout (int/float/tuple): timeout setting for this request. If\n",
      " |              one number provided, it will be total request timeout. It can also\n",
      " |              be a pair (tuple) of (connection, read) timeouts.\n",
      " |              Default is None.\n",
      " |          _check_input_type (bool): specifies if type checking\n",
      " |              should be done one the data sent to the server.\n",
      " |              Default is True.\n",
      " |          _check_return_type (bool): specifies if type checking\n",
      " |              should be done one the data received from the server.\n",
      " |              Default is True.\n",
      " |          _host_index (int/None): specifies the index of the server\n",
      " |              that we want to use.\n",
      " |              Default is read from the configuration.\n",
      " |          async_req (bool): execute request asynchronously\n",
      " |      \n",
      " |      Returns:\n",
      " |          V1AssemblyMetadata\n",
      " |              If the method is called asynchronously, returns the request\n",
      " |              thread.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "genome_api = GenomeApi()\n",
    "help(genome_api)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ae8460a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'assembly': {'annotation_metadata': {'busco': {'busco_lineage': 'eudicots_odb10',\n",
      "                                                'busco_ver': '4.0.2',\n",
      "                                                'complete': 0.99269134,\n",
      "                                                'duplicated': 0.091573514,\n",
      "                                                'fragmented': 0.0017196905,\n",
      "                                                'missing': 0.005588994,\n",
      "                                                'single_copy': 0.9011178,\n",
      "                                                'total_count': '2326'},\n",
      "                                      'file': [{'estimated_size': '11438918',\n",
      "                                                'type': 'GENOME_GFF'},\n",
      "                                               {'estimated_size': '261191555',\n",
      "                                                'type': 'GENOME_GBFF'},\n",
      "                                               {'estimated_size': '23490972',\n",
      "                                                'type': 'RNA_FASTA'},\n",
      "                                               {'estimated_size': '10312194',\n",
      "                                                'type': 'PROT_FASTA'},\n",
      "                                               {'estimated_size': '10506789',\n",
      "                                                'type': 'GENOME_GTF'},\n",
      "                                               {'estimated_size': '16796091',\n",
      "                                                'type': 'CDS_FASTA'}],\n",
      "                                      'name': 'NCBI Juglans regia Annotation '\n",
      "                                              'Release 101',\n",
      "                                      'release_date': '2020-07-27',\n",
      "                                      'release_number': '101',\n",
      "                                      'report_url': 'https://www.ncbi.nlm.nih.gov/genome/annotation_euk/Juglans_regia/101',\n",
      "                                      'source': 'NCBI RefSeq',\n",
      "                                      'stats': {'gene_counts': {'protein_coding': 30709,\n",
      "                                                                'total': 37996}}},\n",
      "              'assembly_accession': 'GCF_001411555.2',\n",
      "              'assembly_category': 'representative genome',\n",
      "              'assembly_level': 'Chromosome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA291087',\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'cultivar:Chandler '\n",
      "                                                                 'Genome '\n",
      "                                                                 'sequencing '\n",
      "                                                                 'and '\n",
      "                                                                 'assembly'}]}],\n",
      "              'biosample': {'accession': 'SAMN03704235',\n",
      "                            'attributes': [{'name': 'cultivar',\n",
      "                                            'value': 'Chandler'},\n",
      "                                           {'name': 'age', 'value': '18'},\n",
      "                                           {'name': 'isolation_source',\n",
      "                                            'value': 'plant at flowering time'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'USA: California, Davis'},\n",
      "                                           {'name': 'tissue',\n",
      "                                            'value': 'leaves'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': 'Plant sample from '\n",
      "                                                     'Juglans regia'},\n",
      "                            'last_updated': '2015-10-05T16:33:37.567',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}], 'name': 'UCDavis'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2015-05-21T17:45:06.623',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': ' Leaves'},\n",
      "                                           {'db': 'SRA', 'value': 'SRS957431'}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2015-05-21T17:45:06.702'},\n",
      "                            'submission_date': '2015-05-21T17:45:06.703'},\n",
      "              'biosample_accession': 'SAMN03704235',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCF_001411555.2',\n",
      "              'chromosomes': [{'accession_version': 'NC_049901.1',\n",
      "                               'gc_count': '16339038',\n",
      "                               'length': '45207397',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'NC_049902.1',\n",
      "                               'gc_count': '13647496',\n",
      "                               'length': '37821870',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'NC_049903.1',\n",
      "                               'gc_count': '12603732',\n",
      "                               'length': '35064427',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'NC_049904.1',\n",
      "                               'gc_count': '12587123',\n",
      "                               'length': '34823025',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'NC_049905.1',\n",
      "                               'gc_count': '8048844',\n",
      "                               'length': '22562875',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'NC_049906.1',\n",
      "                               'gc_count': '14227033',\n",
      "                               'length': '39020271',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'NC_049907.1',\n",
      "                               'gc_count': '18883011',\n",
      "                               'length': '52418484',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'NC_049908.1',\n",
      "                               'gc_count': '10989783',\n",
      "                               'length': '30564197',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'NC_049909.1',\n",
      "                               'gc_count': '8781156',\n",
      "                               'length': '24263475',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'NC_049910.1',\n",
      "                               'gc_count': '13653813',\n",
      "                               'length': '37707155',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'NC_049911.1',\n",
      "                               'gc_count': '13295843',\n",
      "                               'length': '37114715',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'NC_049912.1',\n",
      "                               'gc_count': '11347187',\n",
      "                               'length': '31492331',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'NC_049913.1',\n",
      "                               'gc_count': '14317098',\n",
      "                               'length': '39757759',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'NC_049914.1',\n",
      "                               'gc_count': '10372462',\n",
      "                               'length': '28841373',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'NC_049915.1',\n",
      "                               'gc_count': '7285363',\n",
      "                               'length': '20407330',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'NC_049916.1',\n",
      "                               'gc_count': '10331800',\n",
      "                               'length': '28711772',\n",
      "                               'name': '16'},\n",
      "                              {'accession_version': 'NC_028617.1',\n",
      "                               'gc_count': '57970',\n",
      "                               'length': '160537',\n",
      "                               'name': 'Pltd'},\n",
      "                              {'gc_count': '10479976',\n",
      "                               'length': '27007677',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 1072962,\n",
      "              'display_name': 'Walnut 2.0',\n",
      "              'estimated_size': '695804952',\n",
      "              'gc_count': '207190758',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'cultivar': 'Chandler',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'paired_assembly_accession': 'GCA_001411555.2',\n",
      "              'seq_length': '572786133',\n",
      "              'submission_date': '2020-07-08',\n",
      "              'submitter': 'Johns Hopkins University'}}, {'assembly': {'annotation_metadata': {'file': [{'estimated_size': '7541142',\n",
      "                                                'type': 'GENOME_GFF'},\n",
      "                                               {'estimated_size': '257351110',\n",
      "                                                'type': 'GENOME_GBFF'},\n",
      "                                               {'estimated_size': '8402500',\n",
      "                                                'type': 'PROT_FASTA'},\n",
      "                                               {'estimated_size': '6723254',\n",
      "                                                'type': 'GENOME_GTF'},\n",
      "                                               {'estimated_size': '15044744',\n",
      "                                                'type': 'CDS_FASTA'}],\n",
      "                                      'name': 'Annotation submitted by Johns '\n",
      "                                              'Hopkins University',\n",
      "                                      'release_date': '2020-07-08',\n",
      "                                      'source': 'Johns Hopkins University',\n",
      "                                      'stats': {'gene_counts': {'protein_coding': 37259,\n",
      "                                                                'total': 37265}}},\n",
      "              'assembly_accession': 'GCA_001411555.2',\n",
      "              'assembly_level': 'Chromosome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA291087',\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'cultivar:Chandler '\n",
      "                                                                 'Genome '\n",
      "                                                                 'sequencing '\n",
      "                                                                 'and '\n",
      "                                                                 'assembly'}]}],\n",
      "              'biosample': {'accession': 'SAMN03704235',\n",
      "                            'attributes': [{'name': 'cultivar',\n",
      "                                            'value': 'Chandler'},\n",
      "                                           {'name': 'age', 'value': '18'},\n",
      "                                           {'name': 'isolation_source',\n",
      "                                            'value': 'plant at flowering time'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'USA: California, Davis'},\n",
      "                                           {'name': 'tissue',\n",
      "                                            'value': 'leaves'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': 'Plant sample from '\n",
      "                                                     'Juglans regia'},\n",
      "                            'last_updated': '2015-10-05T16:33:37.567',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}], 'name': 'UCDavis'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2015-05-21T17:45:06.623',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': ' Leaves'},\n",
      "                                           {'db': 'SRA', 'value': 'SRS957431'}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2015-05-21T17:45:06.702'},\n",
      "                            'submission_date': '2015-05-21T17:45:06.703'},\n",
      "              'biosample_accession': 'SAMN03704235',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_001411555.2',\n",
      "              'chromosomes': [{'accession_version': 'CM023841.1',\n",
      "                               'gc_count': '16339038',\n",
      "                               'length': '45207397',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'CM023842.1',\n",
      "                               'gc_count': '13647496',\n",
      "                               'length': '37821870',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'CM023843.1',\n",
      "                               'gc_count': '12603732',\n",
      "                               'length': '35064427',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'CM023844.1',\n",
      "                               'gc_count': '12587123',\n",
      "                               'length': '34823025',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'CM023845.1',\n",
      "                               'gc_count': '8048844',\n",
      "                               'length': '22562875',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'CM023846.1',\n",
      "                               'gc_count': '14227033',\n",
      "                               'length': '39020271',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'CM023847.1',\n",
      "                               'gc_count': '18883011',\n",
      "                               'length': '52418484',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'CM023848.1',\n",
      "                               'gc_count': '10989783',\n",
      "                               'length': '30564197',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'CM023849.1',\n",
      "                               'gc_count': '8781156',\n",
      "                               'length': '24263475',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'CM023850.1',\n",
      "                               'gc_count': '13653813',\n",
      "                               'length': '37707155',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'CM023851.1',\n",
      "                               'gc_count': '13295843',\n",
      "                               'length': '37114715',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'CM023852.1',\n",
      "                               'gc_count': '11347187',\n",
      "                               'length': '31492331',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'CM023853.1',\n",
      "                               'gc_count': '14317098',\n",
      "                               'length': '39757759',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'CM023854.1',\n",
      "                               'gc_count': '10372462',\n",
      "                               'length': '28841373',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'CM023855.1',\n",
      "                               'gc_count': '7285363',\n",
      "                               'length': '20407330',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'CM023856.1',\n",
      "                               'gc_count': '10331800',\n",
      "                               'length': '28711772',\n",
      "                               'name': '16'},\n",
      "                              {'name': 'MT'},\n",
      "                              {'name': 'Pltd'},\n",
      "                              {'gc_count': '10479976',\n",
      "                               'length': '27007677',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 1072962,\n",
      "              'display_name': 'Walnut 2.0',\n",
      "              'estimated_size': '657652854',\n",
      "              'gc_count': '207190758',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'cultivar': 'Chandler',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'paired_assembly_accession': 'GCF_001411555.2',\n",
      "              'seq_length': '572786133',\n",
      "              'submission_date': '2020-07-08',\n",
      "              'submitter': 'Johns Hopkins University'}}, {'assembly': {'assembly_accession': 'GCA_002916465.2',\n",
      "              'assembly_level': 'Chromosome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA356989',\n",
      "                                                        'title': 'Whole genome '\n",
      "                                                                 'sequences of '\n",
      "                                                                 'Juglandaceae '\n",
      "                                                                 'species'}]}],\n",
      "              'biosample': {'accession': 'SAMN08193752',\n",
      "                            'attributes': [{'name': 'isolate', 'value': 'R1'},\n",
      "                                           {'name': 'dev_stage',\n",
      "                                            'value': 'Adult'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'China: Beijing'},\n",
      "                                           {'name': 'tissue', 'value': 'leaf'}],\n",
      "                            'bioprojects': [{'accession': 'PRJNA356989'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': 'Plant sample from '\n",
      "                                                     'Juglans regia'},\n",
      "                            'last_updated': '2018-01-02T14:55:40.433',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}],\n",
      "                                      'name': 'Beijing Normal University'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2017-12-16T00:00:00.000',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': 'Jre_r1'},\n",
      "                                           {'db': 'SRA',\n",
      "                                            'value': 'SRS2762609'}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2017-12-16T22:51:03.566'},\n",
      "                            'submission_date': '2017-12-16T22:51:03.567'},\n",
      "              'biosample_accession': 'SAMN08193752',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_002916465.2',\n",
      "              'chromosomes': [{'accession_version': 'CM036110.1',\n",
      "                               'gc_count': '18248108',\n",
      "                               'length': '50678588',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'CM036111.1',\n",
      "                               'gc_count': '8372181',\n",
      "                               'length': '23436397',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'CM036112.1',\n",
      "                               'gc_count': '15683199',\n",
      "                               'length': '43506672',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'CM036113.1',\n",
      "                               'gc_count': '13414748',\n",
      "                               'length': '37390123',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'CM036114.1',\n",
      "                               'gc_count': '13778034',\n",
      "                               'length': '38468870',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'CM036115.1',\n",
      "                               'gc_count': '10533756',\n",
      "                               'length': '29344405',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'CM036116.1',\n",
      "                               'gc_count': '12807849',\n",
      "                               'length': '35862469',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'CM036117.1',\n",
      "                               'gc_count': '9774881',\n",
      "                               'length': '27399129',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'CM036118.1',\n",
      "                               'gc_count': '13057727',\n",
      "                               'length': '36367524',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'CM036119.1',\n",
      "                               'gc_count': '10023847',\n",
      "                               'length': '27863462',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'CM036120.1',\n",
      "                               'gc_count': '12927856',\n",
      "                               'length': '35942744',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'CM036121.1',\n",
      "                               'gc_count': '8889773',\n",
      "                               'length': '24739769',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'CM036122.1',\n",
      "                               'gc_count': '9959329',\n",
      "                               'length': '27801656',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'CM036123.1',\n",
      "                               'gc_count': '8134775',\n",
      "                               'length': '23015373',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'CM036124.1',\n",
      "                               'gc_count': '12712441',\n",
      "                               'length': '35348395',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'CM036125.1',\n",
      "                               'gc_count': '8191239',\n",
      "                               'length': '23147463',\n",
      "                               'name': '16'},\n",
      "                              {'gc_count': '1925959',\n",
      "                               'length': '4761947',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 3222298,\n",
      "              'display_name': 'ASM291646v2',\n",
      "              'estimated_size': '164806577',\n",
      "              'gc_count': '188435702',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'isolate': 'R1',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'seq_length': '525074986',\n",
      "              'submission_date': '2021-10-27',\n",
      "              'submitter': 'Beijing Normal University'}}, {'assembly': {'assembly_accession': 'GCA_003122785.1',\n",
      "              'assembly_level': 'Scaffold',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA445704',\n",
      "                                                        'title': 'Walnut '\n",
      "                                                                 'Comparative '\n",
      "                                                                 'Genome '\n",
      "                                                                 'Project'}]}],\n",
      "              'biosample': {'accession': 'SAMN08795861',\n",
      "                            'attributes': [{'name': 'isolate',\n",
      "                                            'value': '64-172'},\n",
      "                                           {'name': 'cultivar',\n",
      "                                            'value': 'Chandler'},\n",
      "                                           {'name': 'dev_stage',\n",
      "                                            'value': 'tree'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'USA: UC Davis'},\n",
      "                                           {'name': 'tissue',\n",
      "                                            'value': 'leaves'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': 'Plant sample from '\n",
      "                                                     'Juglans regia'},\n",
      "                            'last_updated': '2018-05-14T09:43:22.421',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}],\n",
      "                                      'name': 'University of California Davis'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2018-03-26T00:00:00.000',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': 'Jreg_Chandler_64-172'}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2018-03-26T12:52:02.679'},\n",
      "                            'submission_date': '2018-03-26T12:52:02.676'},\n",
      "              'biosample_accession': 'SAMN08795861',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_003122785.1',\n",
      "              'chromosomes': [{'gc_count': '222554397',\n",
      "                               'length': '650478281',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 317751,\n",
      "              'display_name': 'ASM312278v1',\n",
      "              'estimated_size': '196458697',\n",
      "              'gc_count': '222554397',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'cultivar': 'Chandler',\n",
      "                      'isolate': '64-172',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'seq_length': '650478281',\n",
      "              'submission_date': '2018-05-14',\n",
      "              'submitter': 'University of California Davis'}}, {'assembly': {'assembly_accession': 'GCA_034508915.1',\n",
      "              'assembly_level': 'Chromosome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA985808',\n",
      "                                                        'parent_accessions': ['PRJNA994954'],\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'haplotype 1 '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'},\n",
      "                                                       {'accession': 'PRJNA994954',\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia genome '\n",
      "                                                                 'sequencing'}]}],\n",
      "              'biosample': {'accession': 'SAMN35811177',\n",
      "                            'attributes': [{'name': 'isolate',\n",
      "                                            'value': 'Donggou'},\n",
      "                                           {'name': 'cultivar',\n",
      "                                            'value': 'Not available'},\n",
      "                                           {'name': 'ecotype',\n",
      "                                            'value': 'Not available'},\n",
      "                                           {'name': 'age', 'value': '2 months'},\n",
      "                                           {'name': 'dev_stage',\n",
      "                                            'value': 'Seedling stage'},\n",
      "                                           {'name': 'collection_date',\n",
      "                                            'value': '2021-11-06'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'China:Xinjiang'},\n",
      "                                           {'name': 'tissue', 'value': 'Leaf'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': 'Plant sample from '\n",
      "                                                     'Juglans regia'},\n",
      "                            'last_updated': '2023-06-20T21:52:04.185',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}],\n",
      "                                      'name': 'Ningxia Academy of Agriculture '\n",
      "                                              'and Forestry Sciences'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2023-06-20T00:00:00.000',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': 'Wild walnut from '\n",
      "                                                     'Gongliu'}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2023-06-20T21:52:04.185'},\n",
      "                            'submission_date': '2023-06-20T21:52:04.185'},\n",
      "              'biosample_accession': 'SAMN35811177',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_034508915.1',\n",
      "              'chromosomes': [{'accession_version': 'CM067457.1',\n",
      "                               'gc_count': '17989205',\n",
      "                               'length': '49766161',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'CM067458.1',\n",
      "                               'gc_count': '10767165',\n",
      "                               'length': '29911514',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'CM067459.1',\n",
      "                               'gc_count': '15470375',\n",
      "                               'length': '42808033',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'CM067460.1',\n",
      "                               'gc_count': '13408863',\n",
      "                               'length': '36997848',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'CM067461.1',\n",
      "                               'gc_count': '13781691',\n",
      "                               'length': '38263849',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'CM067462.1',\n",
      "                               'gc_count': '10218863',\n",
      "                               'length': '28360000',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'CM067463.1',\n",
      "                               'gc_count': '13003043',\n",
      "                               'length': '36178363',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'CM067464.1',\n",
      "                               'gc_count': '10586385',\n",
      "                               'length': '29316000',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'CM067465.1',\n",
      "                               'gc_count': '12267583',\n",
      "                               'length': '34103335',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'CM067466.1',\n",
      "                               'gc_count': '11877770',\n",
      "                               'length': '32769496',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'CM067467.1',\n",
      "                               'gc_count': '14034384',\n",
      "                               'length': '38200000',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'CM067468.1',\n",
      "                               'gc_count': '8933703',\n",
      "                               'length': '24697076',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'CM067469.1',\n",
      "                               'gc_count': '10081011',\n",
      "                               'length': '28039067',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'CM067470.1',\n",
      "                               'gc_count': '8186185',\n",
      "                               'length': '22892000',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'CM067471.1',\n",
      "                               'gc_count': '12646840',\n",
      "                               'length': '34703764',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'CM067472.1',\n",
      "                               'gc_count': '7213209',\n",
      "                               'length': '20194000',\n",
      "                               'name': '16'},\n",
      "                              {'gc_count': '12841385',\n",
      "                               'length': '30470500',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 16857125,\n",
      "              'display_name': 'ASM3450891v1',\n",
      "              'estimated_size': '173683866',\n",
      "              'gc_count': '203307660',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'isolate': 'Donggou',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'seq_length': '557671006',\n",
      "              'submission_date': '2023-12-26',\n",
      "              'submitter': 'Ningxia Academy of Agriculture and Forestry '\n",
      "                           'Sciences'}}, {'assembly': {'assembly_accession': 'GCA_034509015.1',\n",
      "              'assembly_level': 'Complete Genome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA985807',\n",
      "                                                        'parent_accessions': ['PRJNA994954'],\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'haplotype 2 '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'},\n",
      "                                                       {'accession': 'PRJNA994954',\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia genome '\n",
      "                                                                 'sequencing'}]}],\n",
      "              'biosample': {'accession': 'SAMN35811177',\n",
      "                            'attributes': [{'name': 'isolate',\n",
      "                                            'value': 'Donggou'},\n",
      "                                           {'name': 'cultivar',\n",
      "                                            'value': 'Not available'},\n",
      "                                           {'name': 'ecotype',\n",
      "                                            'value': 'Not available'},\n",
      "                                           {'name': 'age', 'value': '2 months'},\n",
      "                                           {'name': 'dev_stage',\n",
      "                                            'value': 'Seedling stage'},\n",
      "                                           {'name': 'collection_date',\n",
      "                                            'value': '2021-11-06'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'China:Xinjiang'},\n",
      "                                           {'name': 'tissue', 'value': 'Leaf'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': 'Plant sample from '\n",
      "                                                     'Juglans regia'},\n",
      "                            'last_updated': '2023-06-20T21:52:04.185',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}],\n",
      "                                      'name': 'Ningxia Academy of Agriculture '\n",
      "                                              'and Forestry Sciences'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2023-06-20T00:00:00.000',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': 'Wild walnut from '\n",
      "                                                     'Gongliu'}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2023-06-20T21:52:04.185'},\n",
      "                            'submission_date': '2023-06-20T21:52:04.185'},\n",
      "              'biosample_accession': 'SAMN35811177',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_034509015.1',\n",
      "              'chromosomes': [{'accession_version': 'CM067426.1',\n",
      "                               'gc_count': '17974389',\n",
      "                               'length': '49731481',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'CM067427.1',\n",
      "                               'gc_count': '10703393',\n",
      "                               'length': '29754658',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'CM067428.1',\n",
      "                               'gc_count': '15782432',\n",
      "                               'length': '43667281',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'CM067429.1',\n",
      "                               'gc_count': '13345105',\n",
      "                               'length': '36810695',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'CM067430.1',\n",
      "                               'gc_count': '13758050',\n",
      "                               'length': '38197910',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'CM067431.1',\n",
      "                               'gc_count': '10218894',\n",
      "                               'length': '28360000',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'CM067432.1',\n",
      "                               'gc_count': '12998262',\n",
      "                               'length': '36168672',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'CM067433.1',\n",
      "                               'gc_count': '10644189',\n",
      "                               'length': '29474921',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'CM067434.1',\n",
      "                               'gc_count': '12198616',\n",
      "                               'length': '33910957',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'CM067435.1',\n",
      "                               'gc_count': '11587615',\n",
      "                               'length': '32024469',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'CM067436.1',\n",
      "                               'gc_count': '14038458',\n",
      "                               'length': '38210000',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'CM067437.1',\n",
      "                               'gc_count': '8827234',\n",
      "                               'length': '24396115',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'CM067438.1',\n",
      "                               'gc_count': '10219186',\n",
      "                               'length': '28418685',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'CM067439.1',\n",
      "                               'gc_count': '7998825',\n",
      "                               'length': '22406140',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'CM067440.1',\n",
      "                               'gc_count': '12643739',\n",
      "                               'length': '34698936',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'CM067441.1',\n",
      "                               'gc_count': '7206185',\n",
      "                               'length': '20170000',\n",
      "                               'name': '16'}],\n",
      "              'contig_n50': 11526675,\n",
      "              'display_name': 'ASM3450901v1',\n",
      "              'estimated_size': '165429551',\n",
      "              'gc_count': '190144572',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'isolate': 'Donggou',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'seq_length': '526400920',\n",
      "              'submission_date': '2023-12-26',\n",
      "              'submitter': 'Ningxia Academy of Agriculture and Forestry '\n",
      "                           'Sciences'}}, {'assembly': {'assembly_accession': 'GCA_041380835.1',\n",
      "              'assembly_level': 'Chromosome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA1128727',\n",
      "                                                        'parent_accessions': ['PRJNA1134227'],\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'collapsed '\n",
      "                                                                 'haploid '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'},\n",
      "                                                       {'accession': 'PRJNA1134227',\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'collapsed '\n",
      "                                                                 'haploid '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'}]}],\n",
      "              'biosample': {'accession': 'SAMN42100985',\n",
      "                            'attributes': [{'name': 'isolate',\n",
      "                                            'value': 'Sharkey'},\n",
      "                                           {'name': 'age', 'value': 'adult'},\n",
      "                                           {'name': 'collection_date',\n",
      "                                            'value': '2022-04'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'USA: Davis, California'},\n",
      "                                           {'name': 'tissue', 'value': 'leaf'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': \"Juglans regia 'Sharkey'\"},\n",
      "                            'last_updated': '2024-06-26T20:35:04.462',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}],\n",
      "                                      'name': 'University of California, '\n",
      "                                              'Davis'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2024-06-26T00:00:00.000',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': 'Juglans regia '\n",
      "                                                     \"'Sharkey'\"}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2024-06-26T20:35:04.462'},\n",
      "                            'submission_date': '2024-06-26T20:35:04.462'},\n",
      "              'biosample_accession': 'SAMN42100985',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_041380835.1',\n",
      "              'chromosomes': [{'accession_version': 'CM084135.1',\n",
      "                               'gc_count': '15882080',\n",
      "                               'length': '43905911',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'CM084136.1',\n",
      "                               'gc_count': '13807266',\n",
      "                               'length': '37801083',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'CM084137.1',\n",
      "                               'gc_count': '12307570',\n",
      "                               'length': '34202718',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'CM084138.1',\n",
      "                               'gc_count': '12565138',\n",
      "                               'length': '34624414',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'CM084139.1',\n",
      "                               'gc_count': '8407046',\n",
      "                               'length': '23474879',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'CM084140.1',\n",
      "                               'gc_count': '13245297',\n",
      "                               'length': '36303274',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'CM084141.1',\n",
      "                               'gc_count': '18464834',\n",
      "                               'length': '50896343',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'CM084142.1',\n",
      "                               'gc_count': '11123577',\n",
      "                               'length': '30578878',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'CM084143.1',\n",
      "                               'gc_count': '8802779',\n",
      "                               'length': '24308908',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'CM084144.1',\n",
      "                               'gc_count': '13661664',\n",
      "                               'length': '37681297',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'CM084145.1',\n",
      "                               'gc_count': '13360342',\n",
      "                               'length': '37128209',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'CM084146.1',\n",
      "                               'gc_count': '10695463',\n",
      "                               'length': '29698517',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'CM084147.1',\n",
      "                               'gc_count': '13927078',\n",
      "                               'length': '38631611',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'CM084148.1',\n",
      "                               'gc_count': '10461975',\n",
      "                               'length': '29044038',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'CM084149.1',\n",
      "                               'gc_count': '6958159',\n",
      "                               'length': '19542513',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'CM084150.1',\n",
      "                               'gc_count': '10355502',\n",
      "                               'length': '28716080',\n",
      "                               'name': '16'},\n",
      "                              {'gc_count': '4040836',\n",
      "                               'length': '8454099',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 21126537,\n",
      "              'display_name': 'ASM4138083v1',\n",
      "              'estimated_size': '170254585',\n",
      "              'gc_count': '198066606',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'isolate': 'Sharkey',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'seq_length': '544992772',\n",
      "              'submission_date': '2024-08-20',\n",
      "              'submitter': 'University of California, Davis'}}, {'assembly': {'assembly_accession': 'GCA_041380855.1',\n",
      "              'assembly_level': 'Chromosome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA1128728',\n",
      "                                                        'parent_accessions': ['PRJNA1134227'],\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'haplotype 2 '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'},\n",
      "                                                       {'accession': 'PRJNA1134227',\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'collapsed '\n",
      "                                                                 'haploid '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'}]}],\n",
      "              'biosample': {'accession': 'SAMN42100985',\n",
      "                            'attributes': [{'name': 'isolate',\n",
      "                                            'value': 'Sharkey'},\n",
      "                                           {'name': 'age', 'value': 'adult'},\n",
      "                                           {'name': 'collection_date',\n",
      "                                            'value': '2022-04'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'USA: Davis, California'},\n",
      "                                           {'name': 'tissue', 'value': 'leaf'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': \"Juglans regia 'Sharkey'\"},\n",
      "                            'last_updated': '2024-06-26T20:35:04.462',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}],\n",
      "                                      'name': 'University of California, '\n",
      "                                              'Davis'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2024-06-26T00:00:00.000',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': 'Juglans regia '\n",
      "                                                     \"'Sharkey'\"}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2024-06-26T20:35:04.462'},\n",
      "                            'submission_date': '2024-06-26T20:35:04.462'},\n",
      "              'biosample_accession': 'SAMN42100985',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_041380855.1',\n",
      "              'chromosomes': [{'accession_version': 'CM084119.1',\n",
      "                               'gc_count': '15909399',\n",
      "                               'length': '43970777',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'CM084120.1',\n",
      "                               'gc_count': '13800954',\n",
      "                               'length': '37759366',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'CM084121.1',\n",
      "                               'gc_count': '12311743',\n",
      "                               'length': '34218580',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'CM084122.1',\n",
      "                               'gc_count': '11332728',\n",
      "                               'length': '31299445',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'CM084123.1',\n",
      "                               'gc_count': '8351328',\n",
      "                               'length': '23343890',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'CM084124.1',\n",
      "                               'gc_count': '12360953',\n",
      "                               'length': '33917405',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'CM084125.1',\n",
      "                               'gc_count': '18180970',\n",
      "                               'length': '50200457',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'CM084126.1',\n",
      "                               'gc_count': '11128164',\n",
      "                               'length': '30591449',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'CM084127.1',\n",
      "                               'gc_count': '8908593',\n",
      "                               'length': '24557279',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'CM084128.1',\n",
      "                               'gc_count': '13241411',\n",
      "                               'length': '36622816',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'CM084129.1',\n",
      "                               'gc_count': '12727986',\n",
      "                               'length': '35453987',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'CM084130.1',\n",
      "                               'gc_count': '10449846',\n",
      "                               'length': '29027587',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'CM084131.1',\n",
      "                               'gc_count': '13696260',\n",
      "                               'length': '37989361',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'CM084132.1',\n",
      "                               'gc_count': '10225662',\n",
      "                               'length': '28404308',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'CM084133.1',\n",
      "                               'gc_count': '6861552',\n",
      "                               'length': '19257108',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'CM084134.1',\n",
      "                               'gc_count': '10236670',\n",
      "                               'length': '28412833',\n",
      "                               'name': '16'},\n",
      "                              {'gc_count': '2143544',\n",
      "                               'length': '4247981',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 13497777,\n",
      "              'display_name': 'ASM4138085v1',\n",
      "              'estimated_size': '165541483',\n",
      "              'gc_count': '191867763',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'isolate': 'Sharkey',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'seq_length': '529274629',\n",
      "              'submission_date': '2024-08-20',\n",
      "              'submitter': 'University of California, Davis'}}, {'assembly': {'assembly_accession': 'GCA_041380885.1',\n",
      "              'assembly_level': 'Chromosome',\n",
      "              'bioproject_lineages': [{'bioprojects': [{'accession': 'PRJNA1128729',\n",
      "                                                        'parent_accessions': ['PRJNA1134227'],\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'haplotype 1 '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'},\n",
      "                                                       {'accession': 'PRJNA1134227',\n",
      "                                                        'title': 'Juglans '\n",
      "                                                                 'regia '\n",
      "                                                                 'collapsed '\n",
      "                                                                 'haploid '\n",
      "                                                                 'genome '\n",
      "                                                                 'sequencing'}]}],\n",
      "              'biosample': {'accession': 'SAMN42100985',\n",
      "                            'attributes': [{'name': 'isolate',\n",
      "                                            'value': 'Sharkey'},\n",
      "                                           {'name': 'age', 'value': 'adult'},\n",
      "                                           {'name': 'collection_date',\n",
      "                                            'value': '2022-04'},\n",
      "                                           {'name': 'geo_loc_name',\n",
      "                                            'value': 'USA: Davis, California'},\n",
      "                                           {'name': 'tissue', 'value': 'leaf'}],\n",
      "                            'description': {'organism': {'organism_name': 'Juglans '\n",
      "                                                                          'regia',\n",
      "                                                         'tax_id': 51240},\n",
      "                                            'title': \"Juglans regia 'Sharkey'\"},\n",
      "                            'last_updated': '2024-06-26T20:35:04.462',\n",
      "                            'models': ['Plant'],\n",
      "                            'owner': {'contacts': [{}],\n",
      "                                      'name': 'University of California, '\n",
      "                                              'Davis'},\n",
      "                            'package': 'Plant.1.0',\n",
      "                            'publication_date': '2024-06-26T00:00:00.000',\n",
      "                            'sample_ids': [{'label': 'Sample name',\n",
      "                                            'value': 'Juglans regia '\n",
      "                                                     \"'Sharkey'\"}],\n",
      "                            'status': {'status': 'live',\n",
      "                                       'when': '2024-06-26T20:35:04.462'},\n",
      "                            'submission_date': '2024-06-26T20:35:04.462'},\n",
      "              'biosample_accession': 'SAMN42100985',\n",
      "              'blast_url': 'https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch&PROG_DEF=blastn&BLAST_SPEC=GDH_GCA_041380885.1',\n",
      "              'chromosomes': [{'accession_version': 'CM084151.1',\n",
      "                               'gc_count': '15857903',\n",
      "                               'length': '43846778',\n",
      "                               'name': '1'},\n",
      "                              {'accession_version': 'CM084152.1',\n",
      "                               'gc_count': '13218773',\n",
      "                               'length': '36536048',\n",
      "                               'name': '2'},\n",
      "                              {'accession_version': 'CM084153.1',\n",
      "                               'gc_count': '12111584',\n",
      "                               'length': '33667154',\n",
      "                               'name': '3'},\n",
      "                              {'accession_version': 'CM084154.1',\n",
      "                               'gc_count': '12959025',\n",
      "                               'length': '35657176',\n",
      "                               'name': '4'},\n",
      "                              {'accession_version': 'CM084155.1',\n",
      "                               'gc_count': '8137363',\n",
      "                               'length': '22750434',\n",
      "                               'name': '5'},\n",
      "                              {'accession_version': 'CM084156.1',\n",
      "                               'gc_count': '13245297',\n",
      "                               'length': '36303274',\n",
      "                               'name': '6'},\n",
      "                              {'accession_version': 'CM084157.1',\n",
      "                               'gc_count': '17994861',\n",
      "                               'length': '49734409',\n",
      "                               'name': '7'},\n",
      "                              {'accession_version': 'CM084158.1',\n",
      "                               'gc_count': '10288879',\n",
      "                               'length': '28485413',\n",
      "                               'name': '8'},\n",
      "                              {'accession_version': 'CM084159.1',\n",
      "                               'gc_count': '8802779',\n",
      "                               'length': '24308908',\n",
      "                               'name': '9'},\n",
      "                              {'accession_version': 'CM084160.1',\n",
      "                               'gc_count': '13424089',\n",
      "                               'length': '37005017',\n",
      "                               'name': '10'},\n",
      "                              {'accession_version': 'CM084161.1',\n",
      "                               'gc_count': '13360342',\n",
      "                               'length': '37128209',\n",
      "                               'name': '11'},\n",
      "                              {'accession_version': 'CM084162.1',\n",
      "                               'gc_count': '10695463',\n",
      "                               'length': '29698517',\n",
      "                               'name': '12'},\n",
      "                              {'accession_version': 'CM084163.1',\n",
      "                               'gc_count': '13948149',\n",
      "                               'length': '38695431',\n",
      "                               'name': '13'},\n",
      "                              {'accession_version': 'CM084164.1',\n",
      "                               'gc_count': '10461975',\n",
      "                               'length': '29044038',\n",
      "                               'name': '14'},\n",
      "                              {'accession_version': 'CM084165.1',\n",
      "                               'gc_count': '6998458',\n",
      "                               'length': '19640399',\n",
      "                               'name': '15'},\n",
      "                              {'accession_version': 'CM084166.1',\n",
      "                               'gc_count': '10363648',\n",
      "                               'length': '28736291',\n",
      "                               'name': '16'},\n",
      "                              {'gc_count': '4819879',\n",
      "                               'length': '10150772',\n",
      "                               'name': 'Un'}],\n",
      "              'contig_n50': 14623224,\n",
      "              'display_name': 'ASM4138088v1',\n",
      "              'estimated_size': '169291040',\n",
      "              'gc_count': '196688467',\n",
      "              'org': {'assembly_counts': {'node': 8, 'subtree': 8},\n",
      "                      'common_name': 'English walnut',\n",
      "                      'isolate': 'Sharkey',\n",
      "                      'parent_tax_id': '16718',\n",
      "                      'rank': 'SPECIES',\n",
      "                      'sci_name': 'Juglans regia',\n",
      "                      'tax_id': '51240'},\n",
      "              'seq_length': '541388268',\n",
      "              'submission_date': '2024-08-20',\n",
      "              'submitter': 'University of California, Davis'}}]\n"
     ]
    }
   ],
   "source": [
    "client = ApiClient()\n",
    "genome_api = GenomeApi(client)\n",
    "\n",
    "# Search for genome assemblies by organism name\n",
    "genome_summaries = genome_api.assembly_descriptors_by_taxon('Juglans regia (English walnut)')\n",
    "print(genome_summaries.assemblies)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cdb4a51",
   "metadata": {},
   "source": [
    "#### Reading Data via Entrez API (affiliated with NCBI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d995b8ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting Bio\n",
      "  Downloading bio-1.7.1-py3-none-any.whl.metadata (5.7 kB)\n",
      "Collecting biopython>=1.80 (from Bio)\n",
      "  Downloading biopython-1.84-cp311-cp311-win_amd64.whl.metadata (13 kB)\n",
      "Collecting gprofiler-official (from Bio)\n",
      "  Downloading gprofiler_official-1.0.0-py3-none-any.whl.metadata (11 kB)\n",
      "Collecting mygene (from Bio)\n",
      "  Downloading mygene-3.2.2-py2.py3-none-any.whl.metadata (10 kB)\n",
      "Requirement already satisfied: pandas in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from Bio) (2.1.1)\n",
      "Collecting pooch (from Bio)\n",
      "  Downloading pooch-1.8.2-py3-none-any.whl.metadata (10 kB)\n",
      "Requirement already satisfied: requests in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from Bio) (2.31.0)\n",
      "Requirement already satisfied: tqdm in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from Bio) (4.64.1)\n",
      "Requirement already satisfied: numpy in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from biopython>=1.80->Bio) (1.26.0)\n",
      "Collecting biothings-client>=0.2.6 (from mygene->Bio)\n",
      "  Downloading biothings_client-0.3.1-py2.py3-none-any.whl.metadata (9.8 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from pandas->Bio) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from pandas->Bio) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from pandas->Bio) (2023.3)\n",
      "Requirement already satisfied: platformdirs>=2.5.0 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from pooch->Bio) (3.10.0)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from pooch->Bio) (21.3)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from requests->Bio) (2.1.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from requests->Bio) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from requests->Bio) (1.26.15)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from requests->Bio) (2022.9.24)\n",
      "Requirement already satisfied: colorama in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from tqdm->Bio) (0.4.6)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from packaging>=20.0->pooch->Bio) (3.0.9)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\cheif\\appdata\\local\\programs\\python\\python311\\lib\\site-packages (from python-dateutil>=2.8.2->pandas->Bio) (1.16.0)\n",
      "Downloading bio-1.7.1-py3-none-any.whl (280 kB)\n",
      "Downloading biopython-1.84-cp311-cp311-win_amd64.whl (2.8 MB)\n",
      "   ---------------------------------------- 0.0/2.8 MB ? eta -:--:--\n",
      "   ------------------------------------- -- 2.6/2.8 MB 13.7 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.8/2.8 MB 10.1 MB/s eta 0:00:00\n",
      "Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
      "Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
      "Downloading pooch-1.8.2-py3-none-any.whl (64 kB)\n",
      "Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n",
      "Installing collected packages: biopython, pooch, gprofiler-official, biothings-client, mygene, Bio\n",
      "Successfully installed Bio-1.7.1 biopython-1.84 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2 pooch-1.8.2\n"
     ]
    }
   ],
   "source": [
    "!pip install Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f9508dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "83141e61",
   "metadata": {},
   "outputs": [
    {
     "ename": "URLError",
     "evalue": "<urlopen error [Errno 11001] getaddrinfo failed>",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mgaierror\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:1348\u001b[0m, in \u001b[0;36mAbstractHTTPHandler.do_open\u001b[1;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[0;32m   1347\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 1348\u001b[0m     \u001b[43mh\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mselector\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1349\u001b[0m \u001b[43m              \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhas_header\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mTransfer-encoding\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1350\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\http\\client.py:1282\u001b[0m, in \u001b[0;36mHTTPConnection.request\u001b[1;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[0;32m   1281\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Send a complete request to the server.\"\"\"\u001b[39;00m\n\u001b[1;32m-> 1282\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\http\\client.py:1328\u001b[0m, in \u001b[0;36mHTTPConnection._send_request\u001b[1;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[0;32m   1327\u001b[0m     body \u001b[38;5;241m=\u001b[39m _encode(body, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbody\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m-> 1328\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mendheaders\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\http\\client.py:1277\u001b[0m, in \u001b[0;36mHTTPConnection.endheaders\u001b[1;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[0;32m   1276\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CannotSendHeader()\n\u001b[1;32m-> 1277\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_output\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmessage_body\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mencode_chunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencode_chunked\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\http\\client.py:1037\u001b[0m, in \u001b[0;36mHTTPConnection._send_output\u001b[1;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[0;32m   1036\u001b[0m \u001b[38;5;28;01mdel\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer[:]\n\u001b[1;32m-> 1037\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmsg\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1039\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m message_body \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m   1040\u001b[0m \n\u001b[0;32m   1041\u001b[0m     \u001b[38;5;66;03m# create a consistent interface to message_body\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\http\\client.py:975\u001b[0m, in \u001b[0;36mHTTPConnection.send\u001b[1;34m(self, data)\u001b[0m\n\u001b[0;32m    974\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mauto_open:\n\u001b[1;32m--> 975\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    976\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\http\\client.py:1447\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1445\u001b[0m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnect to a host on a given (SSL) port.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m-> 1447\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1449\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_tunnel_host:\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\http\\client.py:941\u001b[0m, in \u001b[0;36mHTTPConnection.connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    940\u001b[0m sys\u001b[38;5;241m.\u001b[39maudit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttp.client.connect\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mself\u001b[39m, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mport)\n\u001b[1;32m--> 941\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    942\u001b[0m \u001b[43m    \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msource_address\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    943\u001b[0m \u001b[38;5;66;03m# Might fail in OSs that don't implement TCP_NODELAY\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\socket.py:826\u001b[0m, in \u001b[0;36mcreate_connection\u001b[1;34m(address, timeout, source_address, all_errors)\u001b[0m\n\u001b[0;32m    825\u001b[0m exceptions \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m--> 826\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mSOCK_STREAM\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[0;32m    827\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\socket.py:961\u001b[0m, in \u001b[0;36mgetaddrinfo\u001b[1;34m(host, port, family, type, proto, flags)\u001b[0m\n\u001b[0;32m    960\u001b[0m addrlist \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m--> 961\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43m_socket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mproto\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[0;32m    962\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "\u001b[1;31mgaierror\u001b[0m: [Errno 11001] getaddrinfo failed",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mURLError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[10], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m Entrez\u001b[38;5;241m.\u001b[39memail \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124madamhetherwick@gmail.com\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m----> 2\u001b[0m handle \u001b[38;5;241m=\u001b[39m \u001b[43mEntrez\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43megquery\u001b[49m\u001b[43m(\u001b[49m\u001b[43mterm\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mJuglans regia (English walnut)\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m      3\u001b[0m record \u001b[38;5;241m=\u001b[39m Entrez\u001b[38;5;241m.\u001b[39mread(handle)\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\Entrez\\__init__.py:385\u001b[0m, in \u001b[0;36megquery\u001b[1;34m(**keywds)\u001b[0m\n\u001b[0;32m    383\u001b[0m variables\u001b[38;5;241m.\u001b[39mupdate(keywds)\n\u001b[0;32m    384\u001b[0m request \u001b[38;5;241m=\u001b[39m _build_request(cgi, variables)\n\u001b[1;32m--> 385\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\Entrez\\__init__.py:634\u001b[0m, in \u001b[0;36m_open\u001b[1;34m(request)\u001b[0m\n\u001b[0;32m    632\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(max_tries):\n\u001b[0;32m    633\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 634\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    635\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m HTTPError \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[0;32m    636\u001b[0m         \u001b[38;5;66;03m# Reraise if the final try fails\u001b[39;00m\n\u001b[0;32m    637\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m i \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m max_tries \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:216\u001b[0m, in \u001b[0;36murlopen\u001b[1;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[0;32m    214\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    215\u001b[0m     opener \u001b[38;5;241m=\u001b[39m _opener\n\u001b[1;32m--> 216\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mopener\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:525\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[1;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[0;32m    523\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m processor \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprocess_response\u001b[38;5;241m.\u001b[39mget(protocol, []):\n\u001b[0;32m    524\u001b[0m     meth \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(processor, meth_name)\n\u001b[1;32m--> 525\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mmeth\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresponse\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    527\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:634\u001b[0m, in \u001b[0;36mHTTPErrorProcessor.http_response\u001b[1;34m(self, request, response)\u001b[0m\n\u001b[0;32m    631\u001b[0m \u001b[38;5;66;03m# According to RFC 2616, \"2xx\" code indicates that the client's\u001b[39;00m\n\u001b[0;32m    632\u001b[0m \u001b[38;5;66;03m# request was successfully received, understood, and accepted.\u001b[39;00m\n\u001b[0;32m    633\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;241m200\u001b[39m \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m code \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m300\u001b[39m):\n\u001b[1;32m--> 634\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparent\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43merror\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    635\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mhttp\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresponse\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmsg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhdrs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    637\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:557\u001b[0m, in \u001b[0;36mOpenerDirector.error\u001b[1;34m(self, proto, *args)\u001b[0m\n\u001b[0;32m    555\u001b[0m     http_err \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m    556\u001b[0m args \u001b[38;5;241m=\u001b[39m (\u001b[38;5;28mdict\u001b[39m, proto, meth_name) \u001b[38;5;241m+\u001b[39m args\n\u001b[1;32m--> 557\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_chain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    558\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result:\n\u001b[0;32m    559\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:496\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[1;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[0;32m    494\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[0;32m    495\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[1;32m--> 496\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    497\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    498\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:749\u001b[0m, in \u001b[0;36mHTTPRedirectHandler.http_error_302\u001b[1;34m(self, req, fp, code, msg, headers)\u001b[0m\n\u001b[0;32m    746\u001b[0m fp\u001b[38;5;241m.\u001b[39mread()\n\u001b[0;32m    747\u001b[0m fp\u001b[38;5;241m.\u001b[39mclose()\n\u001b[1;32m--> 749\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparent\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnew\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:519\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[1;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[0;32m    516\u001b[0m     req \u001b[38;5;241m=\u001b[39m meth(req)\n\u001b[0;32m    518\u001b[0m sys\u001b[38;5;241m.\u001b[39maudit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124murllib.Request\u001b[39m\u001b[38;5;124m'\u001b[39m, req\u001b[38;5;241m.\u001b[39mfull_url, req\u001b[38;5;241m.\u001b[39mdata, req\u001b[38;5;241m.\u001b[39mheaders, req\u001b[38;5;241m.\u001b[39mget_method())\n\u001b[1;32m--> 519\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    521\u001b[0m \u001b[38;5;66;03m# post-process response\u001b[39;00m\n\u001b[0;32m    522\u001b[0m meth_name \u001b[38;5;241m=\u001b[39m protocol\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_response\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:536\u001b[0m, in \u001b[0;36mOpenerDirector._open\u001b[1;34m(self, req, data)\u001b[0m\n\u001b[0;32m    533\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[0;32m    535\u001b[0m protocol \u001b[38;5;241m=\u001b[39m req\u001b[38;5;241m.\u001b[39mtype\n\u001b[1;32m--> 536\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_chain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_open\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\n\u001b[0;32m    537\u001b[0m \u001b[43m                          \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_open\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    538\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result:\n\u001b[0;32m    539\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:496\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[1;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[0;32m    494\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[0;32m    495\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[1;32m--> 496\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    497\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    498\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:1391\u001b[0m, in \u001b[0;36mHTTPSHandler.https_open\u001b[1;34m(self, req)\u001b[0m\n\u001b[0;32m   1390\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mhttps_open\u001b[39m(\u001b[38;5;28mself\u001b[39m, req):\n\u001b[1;32m-> 1391\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhttp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mHTTPSConnection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1392\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_context\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_hostname\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\cheif\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\urllib\\request.py:1351\u001b[0m, in \u001b[0;36mAbstractHTTPHandler.do_open\u001b[1;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[0;32m   1348\u001b[0m         h\u001b[38;5;241m.\u001b[39mrequest(req\u001b[38;5;241m.\u001b[39mget_method(), req\u001b[38;5;241m.\u001b[39mselector, req\u001b[38;5;241m.\u001b[39mdata, headers,\n\u001b[0;32m   1349\u001b[0m                   encode_chunked\u001b[38;5;241m=\u001b[39mreq\u001b[38;5;241m.\u001b[39mhas_header(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTransfer-encoding\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[0;32m   1350\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n\u001b[1;32m-> 1351\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m URLError(err)\n\u001b[0;32m   1352\u001b[0m     r \u001b[38;5;241m=\u001b[39m h\u001b[38;5;241m.\u001b[39mgetresponse()\n\u001b[0;32m   1353\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n",
      "\u001b[1;31mURLError\u001b[0m: <urlopen error [Errno 11001] getaddrinfo failed>"
     ]
    }
   ],
   "source": [
    "Entrez.email = \"adamhetherwick@gmail.com\"\n",
    "handle = Entrez.egquery(term=\"Juglans regia (English walnut)\")\n",
    "record = Entrez.read(handle)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
